File: 14_stacks_and_queues_2/05_sliding_window_maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        deque = []
        n = len(nums)
        res = []

        for ind, num in enumerate(nums):
            if deque and deque[0] <= ind - k:
                deque.pop(0)

            while deque and nums[deque[-1]] < num:
                deque.pop()

            deque.append(ind)

            if ind >= k - 1:
                res.append(nums[deque[0]])

        return res

File: 14_stacks_and_queues_2/05_sliding_window_maximum.rs
impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut deque: Vec<usize> = Vec::new();
        let mut res: Vec<i32> = Vec::new();

        for (ind, num) in nums.iter().enumerate() {
            if !deque.is_empty() && deque[0] as i32 <= ind as i32 - k {
                deque.remove(0);
            }

            while !deque.is_empty() && nums[*deque.last().unwrap()] <= *num {
                deque.pop();
            }

            deque.push(ind);
            if ind >= (k - 1) as usize {
                res.push(nums[deque[0]]);
            }
        }

        res
    }
}